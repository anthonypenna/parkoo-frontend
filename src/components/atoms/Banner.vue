<template>
  <transition name="banner">
    <div
      v-if="showBanner"
      v-html="bannerData"
      :class="{
        banner: true,
        'banner--success': success,
        'banner--error': error,
      }"
    />
  </transition>
</template>

<script lang="ts">
import {
  BannerType,
  BANNER_TIMEOUT,
  CLOSED_BANNER_STATE,
} from "@/constants/banner";
import Vue from "vue";
import { mapGetters, mapMutations } from "vuex";

export default Vue.extend({
  name: "Banner",

  computed: {
    ...mapGetters(["showBanner", "bannerType", "bannerData"]),

    success(): boolean {
      return this.bannerType === BannerType.Success;
    },
    error(): boolean {
      return this.bannerType === BannerType.Error;
    },
  },

  methods: {
    ...mapMutations(["setBannerState"]),
  },

  data() {
    return {
      timeoutID: undefined as NodeJS.Timeout | undefined,
    };
  },

  destroyed() {
    clearTimeout(this.timeoutID as NodeJS.Timeout);
  },

  watch: {
    showBanner() {
      if (!this.showBanner) {
        clearTimeout(this.timeoutID as NodeJS.Timeout);
        return;
      }

      this.timeoutID = setTimeout(() => {
        clearTimeout(this.timeoutID as NodeJS.Timeout);
        this.setBannerState(CLOSED_BANNER_STATE);
      }, BANNER_TIMEOUT);
    },
  },
});
</script>

<style lang="scss" scoped>
.banner {
  position: fixed;
  top: 32px;
  left: 0;
  width: 100%;
  padding-top: 1.5rem;
  border-bottom-left-radius: 15px;
  border-bottom-right-radius: 15px;
  color: #fff;
  z-index: 10;

  &--error {
    background: rgba($color: #ff7373, $alpha: 0.6);
    box-shadow: 0 4px 16px rgba($color: #ff7373, $alpha: 0.4);
  }

  &--success {
    background: rgba($color: #1ddc3c, $alpha: 0.6);
    box-shadow: 0 4px 16px rgba($color: #1ddc3c, $alpha: 0.4);
  }
}

.banner-enter-active,
.banner-leave-active {
  transition: transform 1s ease;
}

.banner-enter {
  transform: translateY(-100%);
}

.banner-enter-to {
  transform: translateY(0);
}

.banner-leave {
  transform: translateY(0);
}

.banner-leave-to {
  transform: translateY(-100%);
}
</style>
