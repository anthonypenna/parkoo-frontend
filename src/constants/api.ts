export const API_BASEURL = process.env.VUE_APP_API_BASEURL || ''
