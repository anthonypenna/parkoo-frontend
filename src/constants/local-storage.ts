export const LAST_KNOWN_POSITION_KEY = 'lkp'
