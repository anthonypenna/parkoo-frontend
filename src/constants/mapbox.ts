import { LngLatLike } from 'mapbox-gl'

export const MAPBOX_ACCESS_TOKEN = process.env.VUE_APP_MAPBOX_TOKEN || ''
export const MAPBOX_CENTER = [0, 0] as LngLatLike
export const MAPBOX_THEME = 'mapbox://styles/anthonypenna/ckwt4kriz2ivh15nvz54lzhz3?optimize=true'
export const MAPBOX_ZOOM = 17
export const MAPBOX_FLY_SPEED = 6
