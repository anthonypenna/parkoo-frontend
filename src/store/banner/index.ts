import { actions } from './actions'
import { getters } from './getters'
import { mutations } from './mutations'
import { state } from './state'

export const banner = {
  namespaced: true,
  actions,
  getters,
  mutations,
  state
}
