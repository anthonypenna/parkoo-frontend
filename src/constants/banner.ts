import { BannerState } from '@/models/BannerState'

export const BANNER_TIMEOUT = 3000

export enum BannerType {
  Success,
  Error
}

export const CLOSED_BANNER_STATE = Object.freeze<BannerState>({
  text: '',
  type: BannerType.Error,
  visible: false
})
