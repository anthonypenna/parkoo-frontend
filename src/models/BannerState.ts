import { BannerType } from '@/constants/banner'

export interface BannerState {
  visible: boolean
  type: BannerType
  text: string
}
