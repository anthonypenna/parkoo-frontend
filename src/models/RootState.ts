import { BannerType } from '@/constants/banner'

export interface RootState {
  showLoading: boolean
  hasFetchedGeolocation: boolean
  hasFetchedStreets: boolean
  showNoStreetsModal: boolean
  showAddStreetModal: boolean
  showBanner: boolean
  bannerType: BannerType
  bannerData: string
}
