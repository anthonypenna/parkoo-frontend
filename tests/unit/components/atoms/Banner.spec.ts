import {
  createLocalVue,
  shallowMount,
  ThisTypedShallowMountOptions,
  Wrapper,
} from '@vue/test-utils'
import Banner from '@/components/atoms/Banner.vue'
import { VueConstructor } from 'vue'
import { Store } from 'vuex'
import { RootState } from '@/models/RootState'
import Vuex from 'vuex'
import { BannerType } from '@/constants/banner'
import { BannerState } from '@/models/BannerState'

type BannerWrapper = Wrapper<
  Vue & {
    setBannerState(state: BannerState): void
    timeoutID: NodeJS.Timeout
  }
>

const storeConfig = {
  state: {
    bannerData: '',
    bannerType: BannerType.Error,
    hasFetchedGeolocation: false,
    hasFetchedStreets: false,
    showAddStreetModal: false,
    showBanner: false,
    showLoading: false,
    showNoStreetsModal: false,
  },
  mutations: {
    setBannerState: jest.fn((state, bannerState) => {
      state.bannerData = bannerState.text
      state.bannerType = bannerState.type
      state.showBanner = bannerState.visible
    }),
  },
  getters: {
    showBanner: jest.fn(),
    bannerType: jest.fn(),
    bannerData: jest.fn(),
  },
}

describe('<Banner />', () => {
  let localVue: VueConstructor<Vue>
  let store: Store<RootState>
  let componentOptions: ThisTypedShallowMountOptions<Vue>

  beforeEach(() => {
    localVue = createLocalVue()
    localVue.use(Vuex)
    store = new Store(storeConfig)
    componentOptions = { localVue, store }
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  describe('when hidden', () => {
    beforeEach(() => {
      storeConfig.getters.showBanner.mockReturnValue(false)
    })

    it('should render correctly', () => {
      const wrapper = shallowMount(Banner, componentOptions)
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('when visible', () => {
    beforeEach(() => {
      storeConfig.getters.showBanner.mockReturnValue(true)
      storeConfig.getters.bannerType.mockReturnValue(BannerType.Error)
    })

    it('should render correctly', () => {
      const wrapper = shallowMount(Banner, componentOptions)
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('when success type', () => {
    beforeEach(() => {
      storeConfig.getters.showBanner.mockReturnValue(true)
      storeConfig.getters.bannerType.mockReturnValue(BannerType.Success)
    })

    it('should render correctly', () => {
      const wrapper = shallowMount(Banner, componentOptions)
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('when destroyed', () => {
    beforeEach(() => {
      storeConfig.getters.showBanner.mockReturnValue(false)
    })

    it('should clear internal timeout', () => {
      const clearTimeout = jest.spyOn(global, 'clearTimeout')
      const wrapper = shallowMount(Banner, componentOptions) as BannerWrapper

      wrapper.destroy()
      expect(clearTimeout).toHaveBeenCalledWith(wrapper.vm.timeoutID)
    })
  })
})
