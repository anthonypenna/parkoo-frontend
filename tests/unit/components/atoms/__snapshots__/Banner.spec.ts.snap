// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Banner /> when hidden should render correctly 1`] = `
<transition-stub name="banner">
  <!---->
</transition-stub>
`;

exports[`<Banner /> when success type should render correctly 1`] = `
<transition-stub name="banner">
  <div class="banner banner--success"></div>
</transition-stub>
`;

exports[`<Banner /> when visible should render correctly 1`] = `
<transition-stub name="banner">
  <div class="banner banner--error"></div>
</transition-stub>
`;
